### ISSUE TITLE
CUDA out of memory on RTX 3090 when using default hyperparameters

### ISSUE THREAD

**Issue by ysh-1998** [SAI990323/TALLRec#7]
Hello,

I encountered a CUDA out-of-memory error while using the provided script. I am using an RTX 3090(24G) GPU. I was only able to run the script after setting the `--micro_batch_size` to 32 and each epoch takes approximately 9 hours. I would like to know if the default parameters provided in the script are intended to run smoothly on the RTX 3090 GPU. Additionally, I would appreciate information regarding the expected time per epoch, as I would like to ensure that my replication process is error-free.

Below is my running script modified from `./shell/instruct_7B.sh`, the lines marked with "# modified" is different from the original script. Could you please confirm if there are any issues?

Thank you for your assistance!
```shell
echo $1, $2
seed=$2
output_dir="./lora-alpaca" # modified
base_model="decapoda-research/llama-7b-hf" # modified
train_data="./data/movie/train.json" # modified
val_data="./data/movie/valid.json" # modified
instruction_model="tloen/alpaca-lora-7b" # modified
for lr in 1e-4
do
    for dropout in 0.05
    do
        for sample in 512 # modified
        do
                mkdir -p $output_dir
                echo "lr: $lr, dropout: $dropout , seed: $seed, sample: $sample"
                CUDA_VISIBLE_DEVICES=$1 python -u finetune_rec.py \
                    --base_model $base_model \
                    --train_data_path $train_data \
                    --val_data_path $val_data \
                    --output_dir ${output_dir}_${seed}_${sample} \ # modified
                    --batch_size 128 \
                    --micro_batch_size 32 \ # modified
                    --num_epochs 3 \ # modified
                    --learning_rate $lr \
                    --cutoff_len 512 \
                    --lora_r 8 \
                    --lora_alpha 16\
                    --lora_dropout $dropout \
                    --lora_target_modules '[q_proj,v_proj]' \
                    --train_on_inputs \
                    --group_by_length \
                    --resume_from_checkpoint $instruction_model \
                    --sample $sample \
                    --seed $2
        done
    done
done
```

**Comment by SAI990323** [comment]
We apologize for the ambiguity in the Readme. Please note that our default setting does not cater specifically to the Nvidia 3090 (24GB)
However, we have conducted experiments using the '64 samples' setting and the Nvidia 3090 (24GB) GPU. With this configuration, the training process can be completed in approximately 1 hour. The training time may vary due to different seeds being used.

**Comment by ysh-1998** [comment]
Thank you for your response. I appreciate the information provided. 

After following your suggestion and setting the --sample to 64 and --micro_batch_size to 32, it seems that the script can be completed in approximately 1.5 hours. Considering that setting micro_batch_size to 32 did not fully utilize the available GPU memory (20556 / 24268 MB on 3090), increasing micro_batch_size appropriately should allow the script to finish in approximately one hour, as you mentioned.

Below are my running script and terminal output, are they right? Thank you for your assistance!
```shell
echo $1, $2
seed=$2
output_dir="./lora-alpaca" # modified
base_model="decapoda-research/llama-7b-hf" # modified
train_data="./data/movie/train.json" # modified
val_data="./data/movie/valid.json" # modified
instruction_model="tloen/alpaca-lora-7b" # modified
for lr in 1e-4
do
    for dropout in 0.05
    do
        for sample in 64 # modified
        do
                mkdir -p $output_dir
                echo "lr: $lr, dropout: $dropout , seed: $seed, sample: $sample"
                CUDA_VISIBLE_DEVICES=$1 python -u finetune_rec.py \
                    --base_model $base_model \
                    --train_data_path $train_data \
                    --val_data_path $val_data \
                    --output_dir ${output_dir}_${seed}_${sample} \ # modified
                    --batch_size 128 \
                    --micro_batch_size 32 \ # modified
                    --num_epochs 200 \
                    --learning_rate $lr \
                    --cutoff_len 512 \
                    --lora_r 8 \
                    --lora_alpha 16\
                    --lora_dropout $dropout \
                    --lora_target_modules '[q_proj,v_proj]' \
                    --train_on_inputs \
                    --group_by_length \
                    --resume_from_checkpoint $instruction_model \
                    --sample $sample \
                    --seed $2
        done
    done
done
```
```
lora_B.weight', 'base_model.model.model.layers.29.self_attn.q_proj.lora_A.weight', 'base_model.model.model.layers.29.self_attn.q_proj.lora_B.weight', 'base_model.model.model.layers.29.self_attn.v_proj.lora_A.weight', 'base_model.model.model.layers.29.self_attn.v_proj.lora_B.weight', 'base_model.model.model.layers.30.self_attn.q_proj.lora_A.weight', 'base_model.model.model.layers.30.self_attn.q_proj.lora_B.weight', 'base_model.model.model.layers.30.self_attn.v_proj.lora_A.weight', 'base_model.model.model.layers.30.self_attn.v_proj.lora_B.weight', 'base_model.model.model.layers.31.self_attn.q_proj.lora_A.weight', 'base_model.model.model.layers.31.self_attn.q_proj.lora_B.weight', 'base_model.model.model.layers.31.self_attn.v_proj.lora_A.weight', 'base_model.model.model.layers.31.self_attn.v_proj.lora_B.weight'].
{'train_runtime': 5316.3949, 'train_samples_per_second': 2.408, 'train_steps_per_second': 0.038, 'train_loss': 0.31077302873134616, 'epoch': 200.0}                                                                     
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [1:28:36<00:00, 26.58s/it]

 If there's a warning about missing keys above, please disregard :)
```

**Comment by SAI990323** [comment]
The script looks fine, but if you change the $output_dir$ you may need to change the evaluate code to match your naming format

**Comment by ysh-1998** [comment]
OK! Thanks again!

**Comment by ysh-1998** [comment]
Thank you for your previous assistance. I would like to correct the issue I mentioned earlier. It turns out that the problem was related to my personal experimental environment. 

There was a bug in the Alpaca-Lora official code regarding the PEFT version, which caused the CUDA out-of-memory (OOM) issue when using the default parameters you provided. However, after modifying the PEFT version, I found that the script runs smoothly with the default hyperparameters (i.e., `--micro_batch_size=64`).

In summary, the script you provided is indeed working fine with the correct PEFT version. I apologize for my previous oversight and once again express my gratitude for your assistance.

**Comment by WANGBohaO-jpg** [comment]
> Thank you for your previous assistance. I would like to correct the issue I mentioned earlier. It turns out that the problem was related to my personal experimental environment.
> 
> There was a bug in the Alpaca-Lora official code regarding the PEFT version, which caused the CUDA out-of-memory (OOM) issue when using the default parameters you provided. However, after modifying the PEFT version, I found that the script runs smoothly with the default hyperparameters (i.e., `--micro_batch_size=64`).
> 
> In summary, the script you provided is indeed working fine with the correct PEFT version. I apologize for my previous oversight and once again express my gratitude for your assistance.

Hi, I'm facing the same problem. Could you please let me know which version of 'peft' you are using?

**Comment by ysh-1998** [comment]
> > Thank you for your previous assistance. I would like to correct the issue I mentioned earlier. It turns out that the problem was related to my personal experimental environment.
> > There was a bug in the Alpaca-Lora official code regarding the PEFT version, which caused the CUDA out-of-memory (OOM) issue when using the default parameters you provided. However, after modifying the PEFT version, I found that the script runs smoothly with the default hyperparameters (i.e., `--micro_batch_size=64`).
> > In summary, the script you provided is indeed working fine with the correct PEFT version. I apologize for my previous oversight and once again express my gratitude for your assistance.
> 
> Hi, I'm facing the same problem. Could you please let me know which version of 'peft' you are using?

Sure! I used the 0.3.0 version.
